File: project_euler_unified/problem_146/solution.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef unsigned long long ull;
typedef __int128 lll;

ll mod_mul(ll a, ll b, ll m) {
    return (lll)a * b % m;
}

ll mod_pow(ll base, ll exp, ll mod) {
    ll result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) result = mod_mul(result, base, mod);
        base = mod_mul(base, base, mod);
        exp >>= 1;
    }
    return result;
}

bool miller_rabin(ll n, ll a) {
    if (n % a == 0) return n == a;
    ll d = n - 1;
    int r = 0;
    while (d % 2 == 0) { d /= 2; r++; }
    ll x = mod_pow(a, d, n);
    if (x == 1 || x == n - 1) return true;
    for (int i = 0; i < r - 1; i++) {
        x = mod_mul(x, x, n);
        if (x == n - 1) return true;
    }
    return false;
}

bool is_prime(ll n) {
    if (n < 2) return false;
    if (n < 4) return true;
    if (n % 2 == 0 || n % 3 == 0) return false;
    for (ll a : {2, 3, 5, 7, 11, 13, 17, 19, 23}) {
        if (!miller_rabin(n, a)) return false;
    }
    return true;
}

int main() {
    // Precompute valid residues mod 210
    vector<int> valid;
    for (int r = 0; r < 210; r++) {
        if (r % 2 != 0) continue;
        if (r % 3 == 0) continue;
        if (r % 5 != 0) continue;
        // Check mod 7
        ll r2 = (ll)r * r;
        bool ok = true;
        for (int k : {1, 3, 7, 9, 13, 27}) {
            if ((r2 + k) % 7 == 0) { ok = false; break; }
        }
        if (ok) valid.push_back(r);
    }

    ll total = 0;
    int must_prime[] = {1, 3, 7, 9, 13, 27};
    int must_not_prime[] = {5, 11, 15, 17, 19, 21, 23, 25};

    for (int base = 0; base < 150000000; base += 210) {
        for (int r : valid) {
            ll n = base + r;
            if (n <= 0 || n >= 150000000) continue;
            ll n2 = n * n;

            bool ok = true;
            for (int k : must_prime) {
                if (!is_prime(n2 + k)) { ok = false; break; }
            }
            if (!ok) continue;

            for (int k : must_not_prime) {
                if (is_prime(n2 + k)) { ok = false; break; }
            }
            if (!ok) continue;

            total += n;
        }
    }

    cout << total << endl;
    return 0;
}

File: project_euler_unified/problem_146/solution.py
"""
Problem 146: Investigating a Prime Pattern

Find the sum of all n, 0 < n < 150,000,000, such that
n^2+1, n^2+3, n^2+7, n^2+9, n^2+13, n^2+27 are consecutive primes.
"""

def is_prime_miller_rabin(n):
    """Deterministic Miller-Rabin for n < 3.317e14 (sufficient witnesses)."""
    if n < 2:
        return False
    if n < 4:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    # For n < 3.317 * 10^14, witnesses {2,3,5,7,11,13,17} suffice.
    # Our max is ~(1.5e8)^2 + 27 ~ 2.25e16, so use more witnesses.
    d = n - 1
    r = 0
    while d % 2 == 0:
        d //= 2
        r += 1
    for a in [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]:
        if a >= n:
            continue
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = pow(x, x, n) if False else (x * x) % n
            if x == n - 1:
                break
        else:
            return False
    return True

def solve():
    LIMIT = 150_000_000

    # Precompute valid residues mod 210
    valid_residues = []
    for r in range(210):
        if r % 2 != 0:
            continue
        if r % 3 == 0:
            continue
        if r % 5 != 0:
            continue
        # Check mod 7
        r2 = r * r
        ok = True
        for k in [1, 3, 7, 9, 13, 27]:
            if (r2 + k) % 7 == 0:
                ok = False
                break
        if ok:
            valid_residues.append(r)

    must_be_prime = [1, 3, 7, 9, 13, 27]
    must_not_be_prime = [5, 11, 15, 17, 19, 21, 23, 25]

    total = 0
    for base in range(0, LIMIT, 210):
        for r in valid_residues:
            n = base + r
            if n <= 0 or n >= LIMIT:
                continue

            n2 = n * n

            ok = True
            for k in must_be_prime:
                if not is_prime_miller_rabin(n2 + k):
                    ok = False
                    break
            if not ok:
                continue

            for k in must_not_be_prime:
                if is_prime_miller_rabin(n2 + k):
                    ok = False
                    break
            if not ok:
                continue

            total += n

    print(total)

if __name__ == "__main__":
    solve()
